src/sidebar/components/thread-list-omega.js

```diff
@@ -0,0 +1,246 @@
+import { withServices } from '../util/service-context';
+
+import Thread from './thread';
+
+// When we don't have a real measurement of a thread card's height (yet)
+// from the browser, use this as an approximate value, in pixels.
+const THREAD_DEFAULT_HEIGHT = 200;
+
+/**
+ * Calculate the set of `thread`s that should be rendered by estimating which
+ * of the threads are within or near the viewport.
+ *
+ * @param {Thread[]} threads - List of threads in the order they appear
+ * @param {Object} threadHeights - Map of thread ID to measured height
+ * @param {number} scrollPos - Vertical scroll offset of scrollable container
+ * @param {number} windowHeight -
+ *   Height of the visible area of the scrollable container.
+ */
+function calculateVisibleThreads(
+  threads,
+  threadHeights,
+  scrollPos,
+  windowHeight
+) {
+  // Space above the viewport in pixels which should be considered 'on-screen'
+  // when calculating the set of visible threads
+  const MARGIN_ABOVE = 800;
+  // Same as MARGIN_ABOVE but for the space below the viewport
+  const MARGIN_BELOW = 800;
+  // List of threads which are in or near the viewport and need to
+  // actually be rendered.
+  const visibleThreads = [];
+
+  const calculatedHeights = {
+    offscreenLower: 0,
+    offscreenUpper: 0,
+    total: 0,
+  };
+
+  const threadIsAboveViewport = (threadHeight, heights) =>
+    heights.total + threadHeight < scrollPos - MARGIN_ABOVE;
+  const threadIsVisible = (threadHeight, heights) =>
+    heights.total < scrollPos + windowHeight + MARGIN_BELOW;
+
+  threads.forEach(thread => {
+    const threadHeight = threadHeights[thread.id] || THREAD_DEFAULT_HEIGHT;
+    if (threadIsAboveViewport(threadHeight, calculatedHeights)) {
+      calculatedHeights.offscreenUpper += threadHeight;
+    } else if (threadIsVisible(threadHeight, calculatedHeights)) {
+      visibleThreads.push(thread);
+    } else {
+      // thread is below visible viewport
+      calculatedHeights.offscreenLower += threadHeight;
+    }
+    calculatedHeights.total += threadHeight;
+  });
+
+  return {
+    visibleThreads,
+    offscreenUpperHeight: calculatedHeights.offscreenUpper,
+    offscreenLowerHeight: calculatedHeights.offscreenLower,
+  };
+}
+
+function getScrollContainer() {
+  return document.querySelector('.js-thread-list-scroll-root');
+}
+
+/**
+ * Render a list of threads, but only render those that are in or near the
+ * current browser viewport.
+ */
+function ThreadListOmega({ thread, $rootScope }) {
+  const clearSelection = useStore(store => store.clearSelection);
+
+  // Height of the scrollable container. For the moment, this is the same as
+  // the window height.
+  const [windowHeight, setWindowHeight] = useState(window.innerHeight);
+
+  // Scroll offset of scroll container. This is updated after the scroll
+  // container is scrolled, with debouncing to limit update frequency.
+  const [scrollPosition, setScrollPosition] = useState(
+    getScrollContainer().scrollTop
+  );
```

Comment:
The scroll position and window height are now read from `useState` state rather than being read from the DOM directly. This means that updates are observable to the framework.